File: src/main/java/engineer/carrot/warren/thump/config/ModConfiguration.java
package engineer.carrot.warren.thump.config;

import net.minecraftforge.common.config.Configuration;

import java.io.File;

public class ModConfiguration {
    private Configuration generalConfiguration;
    private Configuration serversConfiguration;

    private static final String GENERAL_NAME = "general.cfg";
    private static final String GENERAL_VERSION = "1";
    private static final String SERVERS_NAME = "servers.cfg";
    private static final String SERVERS_VERSION = "1";

    private EventsConfiguration events;
    private GeneralConfiguration general;
    private ServersConfiguration servers;
    private CommandsConfiguration commands;

    public ModConfiguration() {

    }

    public void initialise(File configDirectory) {
        this.generalConfiguration = new Configuration(new File(configDirectory, GENERAL_NAME), GENERAL_VERSION);
        this.serversConfiguration = new Configuration(new File(configDirectory, SERVERS_NAME), SERVERS_VERSION);
    }

    public void loadAllConfigurations() {
        this.loadGeneralConfiguration();
        this.loadServersConfiguration();
    }

    public void saveAllConfigurations() {
        this.saveGeneralConfiguration();
        this.saveServersConfiguration();
    }

    public void loadGeneralConfiguration() {
        this.generalConfiguration.load();

        this.general = new GeneralConfiguration(this.generalConfiguration);
        this.events = new EventsConfiguration(this.generalConfiguration);
        this.commands = new CommandsConfiguration(this.generalConfiguration);
    }

    public void loadServersConfiguration() {
        this.serversConfiguration.load();

        this.servers = new ServersConfiguration(this.serversConfiguration);
    }

    public void saveGeneralConfiguration() {
        this.generalConfiguration.save();
    }

    public void saveServersConfiguration() {
        this.serversConfiguration.save();
    }

    // Mod configuration
    public GeneralConfiguration getGeneral() {
        return this.general;
    }

    public EventsConfiguration getEvents() {
        return this.events;
    }

    public CommandsConfiguration getCommands() {
        return this.commands;
    }

    // Servers configuration
    public ServersConfiguration getServers() {
        return this.servers;
    }
}

File: src/main/kotlin/engineer/carrot/warren/thump/config/ModConfiguration.kt
package engineer.carrot.warren.thump.config

import net.minecraftforge.common.config.ConfigElement
import net.minecraftforge.common.config.Configuration
import net.minecraftforge.fml.client.config.IConfigElement
import java.io.File

class ModConfiguration {
    lateinit var minecraftGeneralConfiguration: Configuration

    lateinit var events: EventsConfiguration

    // Mod configuration
    lateinit var general: GeneralConfiguration
    lateinit var commands: CommandsConfiguration
    lateinit var formats: FormatsConfiguration

    fun initialise(configDirectory: File) {
        this.minecraftGeneralConfiguration = Configuration(File(configDirectory, GENERAL_NAME), GENERAL_VERSION)
    }

    fun generateConfigElements(): List<IConfigElement> {
        val categories = listOf(GeneralConfiguration.CATEGORY, CommandsConfiguration.CATEGORY, EventsConfiguration.MinecraftEvents.CATEGORY, FormatsConfiguration.MinecraftFormats.CATEGORY)

        return categories.map { ConfigElement(minecraftGeneralConfiguration.getCategory(it)) }
    }

    fun loadAllConfigurations() {
        this.loadGeneralConfiguration()
    }

    fun saveAllConfigurations() {
        this.saveGeneralConfiguration()
    }

    fun loadGeneralConfiguration() {
        this.minecraftGeneralConfiguration.load()

        this.general = GeneralConfiguration(this.minecraftGeneralConfiguration)
        this.events = EventsConfiguration(this.minecraftGeneralConfiguration)
        this.commands = CommandsConfiguration(this.minecraftGeneralConfiguration)
        this.formats = FormatsConfiguration(this.minecraftGeneralConfiguration)
    }

    fun saveGeneralConfiguration() {
        this.minecraftGeneralConfiguration.save()
    }

    companion object {

        private val GENERAL_NAME = "general.cfg"
        private val GENERAL_VERSION = "1"

    }
}
